prow/pod-utils/decorate/podspec.go

```diff
@@ -23,36 +23,211 @@ import (
+}
+
+func VolumeMounts() []string {
+	return []string{LogMountName, CodeMountName, ToolsMountName, GcsCredentialsMountName}
+}
+
+func VolumeMountPaths() []string {
+	return []string{LogMountPath, CodeMountPath, ToolsMountPath, GcsCredentialsMountPath}
+}
+
 // ProwJobToPod converts a ProwJob to a Pod that will run the tests.
 func ProwJobToPod(pj kube.ProwJob, buildID string) (*v1.Pod, error) {
 	if pj.Spec.PodSpec == nil {
 		return nil, fmt.Errorf("prowjob %q lacks a pod spec", pj.Name)
 	}
 
-	env, err := downwardapi.EnvForSpec(downwardapi.NewJobSpec(pj.Spec, buildID, pj.Name))
+	rawEnv, err := downwardapi.EnvForSpec(downwardapi.NewJobSpec(pj.Spec, buildID, pj.Name))
 	if err != nil {
 		return nil, err
 	}
+	env := kubeEnv(rawEnv)
 
 	spec := pj.Spec.PodSpec.DeepCopy()
 	spec.RestartPolicy = "Never"
+	spec.Containers[0].Name = "test"
 
-	for i := range spec.InitContainers {
-		if spec.InitContainers[i].Name == "" {
-			spec.InitContainers[i].Name = fmt.Sprintf("%s-%d", pj.ObjectMeta.Name, i)
+	if pj.Spec.DecorationConfig == nil {
+		spec.Containers[0].Env = append(spec.Containers[0].Env, env...)
+	} else {
+		rawEnv[ArtifactsEnv] = ArtifactsPath
+		logMount := kube.VolumeMount{
+			Name:      LogMountName,
+			MountPath: LogMountPath,
 		}
-		spec.InitContainers[i].Env = append(spec.InitContainers[i].Env, kubeEnv(env)...)
-	}
-	for i := range spec.Containers {
-		if spec.Containers[i].Name == "" {
-			spec.Containers[i].Name = fmt.Sprintf("%s-%d", pj.ObjectMeta.Name, i)
+		logVolume := kube.Volume{
+			Name: LogMountName,
+			VolumeSource: kube.VolumeSource{
+				EmptyDir: &kube.EmptyDirVolumeSource{},
+			},
+		}
+
+		codeMount := kube.VolumeMount{
+			Name:      CodeMountName,
+			MountPath: CodeMountPath,
 		}
-		spec.Containers[i].Env = append(spec.Containers[i].Env, kubeEnv(env)...)
+		codeVolume := kube.Volume{
+			Name: CodeMountName,
+			VolumeSource: kube.VolumeSource{
+				EmptyDir: &kube.EmptyDirVolumeSource{},
+			},
+		}
+
+		toolsMount := kube.VolumeMount{
+			Name:      ToolsMountName,
+			MountPath: ToolsMountPath,
+		}
+		toolsVolume := kube.Volume{
+			Name: ToolsMountName,
+			VolumeSource: kube.VolumeSource{
+				EmptyDir: &kube.EmptyDirVolumeSource{},
+			},
+		}
+
+		gcsCredentialsMount := kube.VolumeMount{
+			Name:      GcsCredentialsMountName,
+			MountPath: GcsCredentialsMountPath,
+		}
+		gcsCredentialsVolume := kube.Volume{
+			Name: GcsCredentialsMountName,
+			VolumeSource: kube.VolumeSource{
+				Secret: &kube.SecretSource{
+					SecretName: pj.Spec.DecorationConfig.GCSCredentialsSecret,
+				},
+			},
+		}
+
+		cloneLog := fmt.Sprintf("%s/clone.json", LogMountPath)
+		var refs []*kube.Refs
+		refs = append(refs, pj.Spec.Refs)
+		refs = append(refs, pj.Spec.ExtraRefs...)
+		cloneConfigEnv, err := clonerefs.Encode(clonerefs.Options{
+			SrcRoot:      CodeMountPath,
+			Log:          cloneLog,
+			GitUserName:  clonerefs.DefaultGitUserName,
+			GitUserEmail: clonerefs.DefaultGitUserEmail,
+			GitRefs:      refs,
```

Comment:
nit: 
```go
GitRefs: append(pj.Spec.ExtraRefs, pj.Spec.Refs)
```

Reply:
Actually that would mutate the storage under `pj.Spec.ExtraRefs` which @cjwagner pointed out which is why I do it this way now :)

Reply:
`append` is not mutating its input slice: https://play.golang.org/p/Wkxv2cKTV3f

Reply:
https://stackoverflow.com/questions/35920534/why-does-append-modify-passed-slice

Reply:
right

>  If the capacity of s is not large enough to fit the additional values, append allocates a new, sufficiently large underlying array that fits both the existing slice elements and the additional values. Otherwise, append re-uses the underlying array.